File: src/main/java/fi/metatavu/metaform/server/metaforms/FieldController.java
package fi.metatavu.metaform.server.metaforms;

import java.time.OffsetDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.StringTokenizer;
import java.util.stream.Collectors;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;

import fi.metatavu.metaform.api.spec.model.Metaform;
import fi.metatavu.metaform.api.spec.model.MetaformField;
import fi.metatavu.metaform.api.spec.model.MetaformFieldType;
import fi.metatavu.metaform.api.spec.model.MetaformSection;
import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.math.NumberUtils;
import org.slf4j.Logger;

import fi.metatavu.metaform.server.persistence.dao.AnyReplyFieldDAO;
import fi.metatavu.metaform.server.persistence.dao.AnyTableReplyFieldRowCellDAO;
import fi.metatavu.metaform.server.persistence.dao.AttachmentReplyFieldItemDAO;
import fi.metatavu.metaform.server.persistence.dao.ListReplyFieldItemDAO;
import fi.metatavu.metaform.server.persistence.dao.TableReplyFieldRowDAO;
import fi.metatavu.metaform.server.persistence.model.AttachmentReplyField;
import fi.metatavu.metaform.server.persistence.model.BooleanReplyField;
import fi.metatavu.metaform.server.persistence.model.ListReplyField;
import fi.metatavu.metaform.server.persistence.model.NumberReplyField;
import fi.metatavu.metaform.server.persistence.model.Reply;
import fi.metatavu.metaform.server.persistence.model.ReplyField;
import fi.metatavu.metaform.server.persistence.model.StringReplyField;
import fi.metatavu.metaform.server.persistence.model.TableReplyField;
import fi.metatavu.metaform.server.persistence.model.TableReplyFieldNumberRowCell;
import fi.metatavu.metaform.server.persistence.model.TableReplyFieldRow;
import fi.metatavu.metaform.server.persistence.model.TableReplyFieldRowCell;
import fi.metatavu.metaform.server.persistence.model.TableReplyFieldStringRowCell;
/**
 * Field controller
 * 
 * @author Antti Leppä
 */
@ApplicationScoped
public class FieldController {
  
  @Inject
  private Logger logger;

  @Inject
  private FieldTypeMapper fieldTypeMapper;

  @Inject
  private AnyReplyFieldDAO anyReplyfieldDAO;

  @Inject
  private ListReplyFieldItemDAO listReplyFieldItemDAO;

  @Inject
  private AttachmentReplyFieldItemDAO attachmentReplyFieldItemDAO;
  
  @Inject
  private TableReplyFieldRowDAO tableReplyFieldRowDAO;
  
  @Inject
  private AnyTableReplyFieldRowCellDAO anyTableReplyFieldRowCellDAO;
  
  /**
   * Parses field filters
   * 
   * @param metaform metaform
   * @param filterList list of filters
   * @return parsed filters
   */
  public FieldFilters parseFilters(Metaform metaform, List<String> filterList) {
    if (filterList == null) {
      return null;
    }
    
    List<String> filterListCombined = new ArrayList<>(filterList.size());
    
    filterList.stream()
      .filter(StringUtils::isNoneEmpty)
      .forEach(filter -> filterListCombined.addAll(Arrays.asList(StringUtils.split(filter, ','))));
    
    List<FieldFilter> filters = filterListCombined.stream()
      .map(filter -> parseFilter(metaform, filter))
      .filter(Objects::nonNull)
      .collect(Collectors.toList());
    
    if (filters == null || filters.isEmpty()) {
      return null;
    }
    
    return new FieldFilters(filters);
  }

  /**
   * Resolves field type for a field name
   * 
   * @param metaformEntity metaform
   * @param name field name
   * @return field type for a field name
   */
  public MetaformFieldType getFieldType(Metaform metaformEntity, String name) {
    MetaformField field = getField(metaformEntity, name);
    if (field == null) {
      return null;
    }
    
    return field.getType();
  }

  /**
   * Returns field value for a reply
   * 
   * @param metaformEntity metaform model
   * @param reply reply
   * @param fieldName field name
   * @param replyFieldMap map containing reply fields
   * @return field value or null if not found
   */
  public Object getFieldValue(Metaform metaformEntity, Reply reply, String fieldName, Map<String, ReplyField> replyFieldMap) {
    if (isMetafield(metaformEntity, fieldName)) {
      return resolveMetaField(fieldName, reply);
    }
    
    ReplyField field = replyFieldMap.get(fieldName);
    if (field == null) {
      return null;
    }
    
    if (field instanceof NumberReplyField) {
      return ((NumberReplyField) field).getValue();
    } else if (field instanceof BooleanReplyField) {
      return ((BooleanReplyField) field).getValue();
    } else if (field instanceof StringReplyField) {
      return ((StringReplyField) field).getValue();
    } else if (field instanceof ListReplyField) {
      return listReplyFieldItemDAO.listItemValuesByField((ListReplyField) field);
    } else if (field instanceof AttachmentReplyField) {
      return attachmentReplyFieldItemDAO.listAttachmentIdsByField((AttachmentReplyField) field);
    } else if (field instanceof TableReplyField) {
      return tableReplyFieldRowDAO.listByField((TableReplyField) field).stream()
        .map(this::getTableRowValue).collect(Collectors.toList());
    } else {
      logger.error("Could not resolve {}", fieldName); 
    }
    
    return null;
  }

  /**
   * Returns table field row as map
   * 
   * @param row row
   * @return table field row as map
   */
  private Map<String, Object> getTableRowValue(TableReplyFieldRow row) {
    return anyTableReplyFieldRowCellDAO.listByRow(row).stream()
      .collect(Collectors.toMap(TableReplyFieldRowCell::getName, cell -> {
        if (cell instanceof TableReplyFieldNumberRowCell) {
          return ((TableReplyFieldNumberRowCell) cell).getValue();
        } else if (cell instanceof TableReplyFieldStringRowCell) {
          return ((TableReplyFieldStringRowCell) cell).getValue();
        }
        
        return null;
      }));
  }
  
  /**
   * Returns whether form field is a meta field
   * 
   * @param metaformEntity form
   * @param name name
   * @return whether form field is a meta field
   */
  private boolean isMetafield(Metaform metaformEntity, String name) {
    MetaformField field = getField(metaformEntity, name);
    return field != null && field.getContexts() != null && field.getContexts().contains("META");
  }
  
  /**
   * Returns field by name
   * 
   * @param metaformEntity form
   * @param name name
   * @return field
   */
  public MetaformField getField(Metaform metaformEntity, String name) {
    List<MetaformSection> sections = metaformEntity.getSections();
    
    for (MetaformSection section : sections) {
      for (MetaformField field : section.getFields()) {
        if (name.equals(field.getName())) {
          return field;
        }
      }
    }
    
    return null;
  }
  
  /**
   * Resolves meta field
   * 
   * @param fieldName field name
   * @param entity reply
   * @return meta field value
   */
  public Object resolveMetaField(String fieldName, fi.metatavu.metaform.server.persistence.model.Reply entity) {
    switch (fieldName) {
      case "lastEditor":
        return entity.getUserId();
      case "created":
      case "createdAt":
        return formatDateTime(entity.getCreatedAt());
      case "modified":
        return formatDateTime(entity.getModifiedAt());
      default:
        logger.warn("Metafield {} not recognized", fieldName);
      break;
    }
    
    return null;
  }
  
  /**
   * Returns field name <> type map from Metaform
   * 
   * @param metaformEntity Metaform REST entity
   * @return field name <> type map
   */
  public Map<String, MetaformField> getFieldMap(Metaform metaformEntity) {
    return metaformEntity.getSections().stream()
      .map(MetaformSection::getFields)
      .flatMap(List::stream)
      .collect(Collectors.toMap(MetaformField::getName, field -> field));
  }

  /**
   * Lists field name by type
   * 
   * @param metaformEntity Metaform
   * @param type type
   * @return field names by type
   */
  public List<String> getFieldNamesByType(Metaform metaformEntity, MetaformFieldType type) {
    return metaformEntity.getSections().stream()
      .map(MetaformSection::getFields)
      .flatMap(List::stream)
      .filter(field -> type.equals(field.getType()))
      .map(MetaformField::getName)
      .collect(Collectors.toList());
  }

  /**
   * Returns map of reply fields where reply field name is a key and the field value
   * 
   * @param reply reply
   * @return map of reply fields where reply field name is a key and the field value
   */
  public Map<String, ReplyField> getReplyFieldMap(Reply reply) {
    return anyReplyfieldDAO.listByReply(reply).stream().collect(Collectors.toMap(ReplyField::getName, replyField -> replyField));
  }

  /**
   * Parses a field filter
   * 
   * @param metaformEntity
   * @param filter
   * @return
   */
  private FieldFilter parseFilter(Metaform metaformEntity, String filter) {
    List<String> tokenizedFilter = tokenizeFilter(filter);
    if (tokenizedFilter.size() == 3) {
      String fieldName = tokenizedFilter.get(0);
      String operatorString = tokenizedFilter.get(1);
      String valueString = tokenizedFilter.get(2);
      MetaformFieldType fieldType = getFieldType(metaformEntity, fieldName);
      StoreDataType storeDataType = fieldTypeMapper.getStoreDataType(fieldType);
      FieldFilterOperator operator = null;
      
      if (":".equals(operatorString)) {
        operator = FieldFilterOperator.EQUALS;
      } else  if ("^".equals(operatorString)) {
        operator = FieldFilterOperator.NOT_EQUALS;
      } else { 
        logger.error("Could not parse operator string {}", operatorString);
        return null;
      }
      
      if (storeDataType == null || storeDataType == StoreDataType.NONE) {
        return null;
      }
      
      return new FieldFilter(storeDataType, fieldName, getFieldFilterValue(storeDataType, valueString), operator);
    }
    
    return null;
  }
  
  /**
   * Tokenizes a field filter
   * 
   * @param filter filter string
   * @return tokenized filter
   */
  private List<String> tokenizeFilter(String filter) {
    StringTokenizer stringTokenizer = new StringTokenizer(filter, ":^", true);
    List<String> result = new ArrayList<>(3);
    
    while (stringTokenizer.hasMoreTokens()) {
      result.add(stringTokenizer.nextToken());
    }
    
    return result;
  }
  
  /**
   * Resolves field filter value
   * 
   * @param storeDataType store data type
   * @param valueString value as string
   * @return field filter value
   */
  private Object getFieldFilterValue(StoreDataType storeDataType, String valueString) {
    if (valueString == null) {
      return null;
    }
    
    switch (storeDataType) {
      case BOOLEAN:
        return BooleanUtils.toBooleanObject(valueString);
      case LIST:
      case STRING:
        return valueString;
      case NUMBER:
        try {
          return NumberUtils.createDouble(valueString);
        } catch (NumberFormatException e) {
          return null;
        }
      default:
        logger.error("Failed to parse valueString {}", valueString);
    }
    
    return null;
  }
  
  /**
   * Formats date time in ISO date-time format
   * 
   * @param dateTime date time
   * @return date time in ISO date-time format
   */
  private String formatDateTime(OffsetDateTime dateTime) {
    if (dateTime == null) {
      return null;
    }
    
    return dateTime.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
  }
  
}

File: src/main/kotlin/fi/metatavu/metaform/server/controllers/FieldController.kt
package fi.metatavu.metaform.server.controllers

import fi.metatavu.metaform.api.spec.model.Metaform
import fi.metatavu.metaform.api.spec.model.MetaformField
import fi.metatavu.metaform.api.spec.model.MetaformFieldType
import fi.metatavu.metaform.server.metaform.*
import fi.metatavu.metaform.server.utils.MetaformUtils
import fi.metatavu.metaform.server.persistence.dao.*
import fi.metatavu.metaform.server.persistence.model.*
import org.apache.commons.lang3.BooleanUtils
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.math.NumberUtils
import org.slf4j.Logger
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Controller for Fields
 */
@ApplicationScoped
class FieldController {

    @Inject
    lateinit var logger: Logger

    @Inject
    lateinit var fieldTypeMapper: FieldTypeMapper

    @Inject
    lateinit var anyReplyFieldDAO: AnyReplyFieldDAO

    @Inject
    lateinit var listReplyFieldItemDAO: ListReplyFieldItemDAO

    @Inject
    lateinit var attachmentReplyFieldItemDAO: AttachmentReplyFieldItemDAO

    @Inject
    lateinit var tableReplyFieldRowDAO: TableReplyFieldRowDAO

    @Inject
    lateinit var anyTableReplyFieldRowCellDAO: AnyTableReplyFieldRowCellDAO

    /**
     * Parses field filters
     *
     * @param metaform metaform
     * @param filterList list of filters
     * @return parsed filters
     */
    fun parseFilters(metaform: Metaform, filterList: List<String?>?): FieldFilters? {
        filterList ?: return null
        
        val filterListCombined = filterList
                .filter(StringUtils::isNotEmpty)
                .map { filter: String? -> listOf(*StringUtils.split(filter, ',')) }
                .flatMap { it.toList() }

        val filters = filterListCombined
                .mapNotNull { filter: String -> parseFilter(metaform, filter) }

        return if (filters.isEmpty()) {
            null
        } else FieldFilters(filters)
    }

    /**
     * Resolves field type for a field name
     *
     * @param metaformEntity metaform
     * @param name field name
     * @return field type for a field name
     */
    fun getFieldType(metaformEntity: Metaform, name: String): MetaformFieldType? {
        return getField(metaformEntity, name)?.type
    }

    /**
     * Returns field value for a reply
     *
     * @param metaformEntity metaform model
     * @param reply reply
     * @param fieldName field name
     * @param replyFieldMap map containing reply fields
     * @return field value or null if not found
     */
    fun getFieldValue(metaformEntity: Metaform, reply: Reply, fieldName: String, replyFieldMap: Map<String, ReplyField?>): Any? {
        if (isMetaField(metaformEntity, fieldName)) {
            return resolveMetaField(fieldName, reply)
        }
        val field = replyFieldMap[fieldName] ?: return null

        return when (field) {
            is NumberReplyField -> field.value
            is BooleanReplyField -> field.value
            is StringReplyField -> field.value
            is ListReplyField -> listReplyFieldItemDAO.listItemValuesByField(field)
            is AttachmentReplyField -> attachmentReplyFieldItemDAO.listAttachmentIdsByField(field)
            is TableReplyField -> tableReplyFieldRowDAO.listByField(field).map{ getTableRowValue(it) }
            else -> {
                logger.error("Could not resolve {}", fieldName)
                null
            }
        }
    }

    /**
     * Returns table field row as map
     *
     * @param row row
     * @return table field row as map
     */
    private fun getTableRowValue(row: TableReplyFieldRow): Map<String?, Any?> {
        return anyTableReplyFieldRowCellDAO.listByRow(row).associate { cell ->
            cell.name to when (cell) {
                is TableReplyFieldNumberRowCell -> cell.value
                is TableReplyFieldStringRowCell -> cell.value
                else -> null
            }
        }
    }

    /**
     * Returns whether form field is a meta field
     *
     * @param metaformEntity form
     * @param name name
     * @return whether form field is a meta field
     */
    private fun isMetaField(metaformEntity: Metaform, name: String): Boolean {
        val field = getField(metaformEntity, name)
        return field?.contexts != null && field.contexts.contains("META")
    }

    /**
     * Returns field by name
     *
     * @param metaformEntity form
     * @param name name
     * @return field
     */
    fun getField(metaformEntity: Metaform, name: String): MetaformField? {
        return MetaformUtils.getMetaformFields(metaformEntity).find{ field -> field.name == name }
    }

    /**
     * Resolves meta field
     *
     * @param fieldName field name
     * @param entity reply
     * @return meta field value
     */
    fun resolveMetaField(fieldName: String?, entity: Reply): Any? {
        return when (fieldName) {
            "lastEditor" -> entity.userId
            "created", "createdAt" -> formatDateTime(entity.createdAt)
            "modified" -> formatDateTime(entity.modifiedAt)
            else -> {
                logger.warn("Metafield {} not recognized", fieldName)
                null
            }
        }
    }

    /**
     * Returns field name <> type map from Metaform
     *
     * @param metaformEntity Metaform REST entity
     * @return field name <> type map
     */
    fun getFieldMap(metaformEntity: Metaform): Map<String, MetaformField> {
        return MetaformUtils.getMetaformFields(metaformEntity).associateBy { field -> field.name!! }
    }

    /**
     * Lists field name by type
     *
     * @param metaformEntity Metaform
     * @param type type
     * @return field names by type
     */
    fun getFieldNamesByType(metaformEntity: Metaform, type: MetaformFieldType): List<String> {
        return MetaformUtils.getMetaformFields(metaformEntity)
                .filter { type == it.type }
                .mapNotNull(MetaformField::name)
    }

    /**
     * Returns map of reply fields where reply field name is a key and the field value
     *
     * @param reply reply
     * @return map of reply fields where reply field name is a key and the field value
     */
    fun getReplyFieldMap(reply: Reply): Map<String, ReplyField?> {
        return anyReplyFieldDAO.listByReply(reply).associate { replyField -> replyField.name to replyField }
    }

    /**
     * Parses a field filter
     *
     * @param metaformEntity
     * @param filter
     * @return
     */
    private fun parseFilter(metaformEntity: Metaform, filter: String): FieldFilter? {
        val tokenizedFilter = tokenizeFilter(filter)
        if (tokenizedFilter.size == 3) {
            val fieldName = tokenizedFilter[0]
            val operatorString = tokenizedFilter[1]
            val valueString = tokenizedFilter[2]
            val fieldType = getFieldType(metaformEntity, fieldName)
            val storeDataType = fieldTypeMapper.getStoreDataType(fieldType)
            val operator = when (operatorString) {
                ":" -> FieldFilterOperator.EQUALS
                "^" -> FieldFilterOperator.NOT_EQUALS
                else -> {
                    logger.error("Could not parse operator string {}", operatorString)
                    return null
                }
            }
            return if (storeDataType == null || storeDataType == StoreDataType.NONE) {
                null
            } else FieldFilter(fieldName, getFieldFilterValue(storeDataType, valueString), storeDataType, operator)
        }
        return null
    }

    /**
     * Tokenizes a field filter
     *
     * @param filter filter string
     * @return tokenized filter
     */
    private fun tokenizeFilter(filter: String): List<String> {
        val stringTokenizer = StringTokenizer(filter, ":^", true)
        val result: MutableList<String> = ArrayList(3)
        while (stringTokenizer.hasMoreTokens()) {
            result.add(stringTokenizer.nextToken())
        }
        return result
    }

    /**
     * Resolves field filter value
     *
     * @param storeDataType store data type
     * @param valueString value as string
     * @return field filter value
     */
    private fun getFieldFilterValue(storeDataType: StoreDataType, valueString: String?): Any? {
        if (valueString == null) {
            return null
        }
        return when (storeDataType) {
            StoreDataType.BOOLEAN -> BooleanUtils.toBooleanObject(valueString)
            StoreDataType.LIST, StoreDataType.STRING -> valueString
            StoreDataType.NUMBER -> {
                try {
                    NumberUtils.createDouble(valueString)
                } catch (e: NumberFormatException) {
                    logger.error("Failed to parse valueString {}", valueString)
                    null
                }

            }
            else -> {
                logger.error("Failed to parse valueString {}", valueString)
                null
            }
        }
    }

    /**
     * Formats date time in ISO date-time format
     *
     * @param dateTime date time
     * @return date time in ISO date-time format
     */
    private fun formatDateTime(dateTime: OffsetDateTime?): String? {
        return dateTime?.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
    }

}